File: csharp/proverb/Proverb.cs
using System.Collections.Generic;
using System.Linq;

public static class Proverb
{
    public static string[] Recite(string[] subjects) => MakeProverb(subjects).ToArray();

    private static IEnumerable<string> MakeProverb(IEnumerable<string> subjects)
    {
        if (!subjects.Any())
        {
            yield break;
        }

        IEnumerable<string> temp = subjects;
        IEnumerable<string> sentence;

        while ((sentence = temp.Take(2)).Count() == 2)
        {
            yield return $"For want of a {sentence.First()} the {sentence.Last()} was lost.";
            temp = temp.Skip(1);
        }

        yield return $"And all for the want of a {subjects.First()}.";
    }
}

File: go/proverb/proverb.go
// Package proverb provides a proverbial sequence of sentences given a list of items.
package proverb

import (
	"fmt"
)

// Proverb generates a sequence of proverbial sentences using the given items.
func Proverb(rhyme []string) []string {
	result := []string{}
	count := len(rhyme)

	for index := 0; index < count; index++ {
		var sentence string

		if index == count-1 {
			sentence = fmt.Sprintf("And all for the want of a %s.", rhyme[0])
		} else {
			item1 := rhyme[index]
			item2 := rhyme[index+1]
			sentence = fmt.Sprintf("For want of a %s the %s was lost.", item1, item2)
		}

		result = append(result, sentence)
	}

	return result
}
